File: moedict-android/app/src/main/java/com/yichiuan/moedict/data/MoeRepository.java
package com.yichiuan.moedict.data;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.VisibleForTesting;

import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Formatter;

import io.reactivex.Completable;
import moe.Dictionary;
import moe.Index;
import moe.Word;

public class MoeRepository {

    private static final int MOD_MASK = 0x3FF; // 0b1111111111
    private static final int BUFFER_SIZE = (int)(1024 * 1024 * 1.2); // 1.2 MB

    private static final String MOE_INDEX_DATAPATH = "moe/index.bin";

    private Context context;

    private StringBuilder appendable = new StringBuilder();
    private Formatter formatter = new Formatter(appendable);

    private byte[] dictBuffer;

    private Index index;

    // Private constructor prevents instantiation from other classes
    MoeRepository(Context context) {
        this.context = context;
    }

    public Word getMoeWord(String word) throws IOException {

        int firstWord = word.codePointAt(0);
        int mod = firstWord & MOD_MASK; // code point % 1024

        appendable.setLength(0);
        String dataPath = formatter.format("moe/%d.bin", mod).toString();

        Dictionary dict = Dictionary.getRootAsDictionary(loadDictData(dataPath));
        return dict.wordsByKey(word);
    }

    @VisibleForTesting
    Dictionary getMoeDictionary(int mod) throws IOException {
        appendable.setLength(0);
        String dataPath = formatter.format("moe/%d.bin", mod).toString();
        return Dictionary.getRootAsDictionary(loadDictData(dataPath));
    }

    public Completable loadIndexData() {
        return Completable.fromAction(() -> {
            if (index == null) {
                index = Index.getRootAsIndex(loadData(MOE_INDEX_DATAPATH, null));
            }
        });
    }

    private ByteBuffer loadDictData(String dataPath) throws IOException {
        if (dictBuffer == null) {
            dictBuffer = new byte[BUFFER_SIZE];
        }
        return loadData(dataPath, dictBuffer);
    }

    private ByteBuffer loadData(@NonNull String dataPath, @Nullable byte[] buffer) throws IOException {

        InputStream is = null;
        ByteBuffer byteBuffer = null;
        try {
            is = context.getAssets().open(dataPath);
            byteBuffer = loadData(is, buffer);
        } finally {
            if (is != null) {
                try {
                    is.close();
                } catch (IOException e) {

                }
            }
        }
        return byteBuffer;
    }

    private ByteBuffer loadData(@NonNull InputStream is, @Nullable byte[] buffer) throws IOException {

        int readSize = 0;
        byte[] readBuffer = buffer;

        int size = is.available();

        if (readBuffer == null || readBuffer.length < size) {
            readBuffer = new byte[size];
        }
        readSize = is.read(readBuffer, 0, size);

        is.close();

        return ByteBuffer.wrap(readBuffer, 0, readSize);
    }

    public ArrayList<Integer> search(String query) {
        return index.search(query);
    }

    public String getWord(int position) {
        return index.words(position);
    }
}

File: moedict-android/app/src/main/java/com/yichiuan/moedict/data/MoeRepository.kt
package com.yichiuan.moedict.data

import android.content.Context
import android.support.annotation.VisibleForTesting
import io.reactivex.Completable
import moe.Dictionary
import moe.Index
import moe.Word
import java.io.InputStream
import java.nio.ByteBuffer
import java.util.*

class MoeRepository internal constructor(private val context: Context) {

    companion object {
        private const val MOD_MASK = 0x3FF // 0b1111111111
        private const val BUFFER_SIZE = (1024.0 * 1024.0 * 1.2).toInt() // 1.2 MB
        private const val MOE_INDEX_DATAPATH = "moe/index.bin"
    }

    private val appendable = StringBuilder()
    private val formatter = Formatter(appendable)

    private var dictBuffer: ByteArray? = null

    private var index: Index? = null

    fun getMoeWord(word: String): Word? {

        val firstWord = word.codePointAt(0)
        val mod = firstWord and MOD_MASK // code point % 1024

        appendable.setLength(0)
        val dataPath = formatter.format("moe/%d.bin", mod).toString()

        val dict = Dictionary.getRootAsDictionary(loadDictData(dataPath))
        return dict.wordsByKey(word)
    }

    @VisibleForTesting
    internal fun getMoeDictionary(mod: Int): Dictionary {
        appendable.setLength(0)
        val dataPath = formatter.format("moe/%d.bin", mod).toString()
        return Dictionary.getRootAsDictionary(loadDictData(dataPath))
    }

    fun loadIndexData(): Completable {
        return Completable.fromAction {
            if (index == null) {
                index = Index.getRootAsIndex(loadData(MOE_INDEX_DATAPATH, null))
            }
        }
    }

    private fun loadDictData(dataPath: String): ByteBuffer? {
        if (dictBuffer == null) {
            dictBuffer = ByteArray(BUFFER_SIZE)
        }
        return loadData(dataPath, dictBuffer)
    }


    private fun loadData(dataPath: String, buffer: ByteArray?): ByteBuffer? {

        var byteBuffer: ByteBuffer? = null

        context.assets.open(dataPath).use {
            byteBuffer = loadData(it, buffer)
        }

        return byteBuffer
    }


    private fun loadData(inputStream: InputStream, buffer: ByteArray?): ByteBuffer {

        var readBuffer = buffer

        val size = inputStream.available()

        if (readBuffer == null || readBuffer.size < size) {
            readBuffer = ByteArray(size)
        }
        val readSize = inputStream.read(readBuffer, 0, size)

        return ByteBuffer.wrap(readBuffer, 0, readSize)
    }

    fun search(query: String): ArrayList<Int> {
        return index!!.search(query)
    }

    fun getWord(position: Int): String? {
        return index!!.words(position)
    }
}